src/olmo_core/train/train_module/transformer/train_module.py

```diff
@@ -344,22 +344,31 @@ def train_batch(self, batch: Dict[str, Any], dry_run: bool = False):
         if "labels" not in batch:
             batch["labels"] = get_labels(batch, label_ignore_index=self.label_ignore_index)
 
-        # Record how many instances are going to be skipped (masked out).
-        if (instance_mask := batch.get("instance_mask")) is not None and not dry_run:
-            self.record_metric(
-                "train/masked instances (%)", (~instance_mask).float().mean(), ReduceType.mean
-            )
-
-        # Calculate and record how many tokens are going to be used in the loss.
+        # Calculate how many tokens will be used in the loss.
         batch_num_tokens = batch["labels"].numel()
+        batch_num_tokens_per_instance = batch["labels"].shape[1]
         batch_num_tokens_for_loss = move_to_device(
             (batch["labels"] != self.label_ignore_index).sum(), self.device
         )
-        self.record_metric(
-            "train/masked labels (%)",
-            (batch_num_tokens - batch_num_tokens_for_loss) / batch_num_tokens,
-            ReduceType.mean,
-        )
+
+        if not dry_run:
+            # Record percentage of masked labels.
+            self.record_metric(
+                "train/masked labels (%)",
+                (batch_num_tokens - batch_num_tokens_for_loss) / batch_num_tokens,
+                ReduceType.mean,
+            )
+
+            # Record percentage of masked instances.
+            if (instance_mask := batch.get("instance_mask")) is not None:
+                self.record_metric(
+                    "train/masked instances (%)", (~instance_mask).float().mean(), ReduceType.mean
+                )
+                # WARN: When we mask out instances with the instance filter, we count those tokens
+                # for the loss anyways. They will count as tokens with a zero loss. This means we
+                # get an artificially *low* loss for these batches. But it is really hard (and slow)
+                # to do this properly in a distributed setup.
+                batch_num_tokens_for_loss += (~instance_mask).sum() * batch_num_tokens_per_instance
```

Comment:
@dirkgr thoughts on just adding 1 token per masked instance instead of `batch_num_tokens_per_instance` to avoid making the loss more artificially low than it needs to be?

Reply:
Or a small epsilon. Doesn't have to be an integer.

Reply:
Done